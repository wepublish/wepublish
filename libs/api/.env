DATABASE_URL=postgresql://postgres@localhost:5432/wepublish_test?schema=public