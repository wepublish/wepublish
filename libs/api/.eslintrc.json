{
  "extends": ["../../.eslintrc.json"],
  "ignorePatterns": ["!**/*", "__tests__/api/**/*.ts", "prisma/seed.ts"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    }
  ]
}
