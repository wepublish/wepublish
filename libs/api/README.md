# api

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test api` to execute the unit tests via [Jest](https://jestjs.io).

## Running lint

Run `nx lint api` to execute the lint via [ESLint](https://eslint.org/).
