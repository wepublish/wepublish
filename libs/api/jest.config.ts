/* eslint-disable */
export default {
  displayName: 'api',
  preset: '../../jest.preset.js',
  globals: {},
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]sx?$': [
      'ts-jest',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json'
      }
    ]
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/api',
  setupFilesAfterEnv: ['./__tests__/setup.ts'],
  globalSetup: '<rootDir>/__tests__/setup-database.js',
  collectCoverageFrom: ['!**/*']
}
