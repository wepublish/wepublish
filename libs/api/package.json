{
  "name": "@wepublish/api",
  "version": "9.2.0",
  "description": "API core for we.publish.",
  "keywords": [
    "api",
    "publishing",
    "media",
    "newspaper",
    "cms",
    "subscription",
    "graphql"
  ],
  "homepage": "https://github.com/wepublish/wepublish",
  "bugs": {
    "url": "https://github.com/wepublish/wepublish/issues",
    "email": "development@wepublish.ch"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/wepublish/wepublish.git",
    "directory": "libs/api"
  },
  "license": "MIT",
  "prisma": {
    "schema": "prisma/schema.prisma",
    "seed": "npx ts-node __tests__/seed.ts"
  },
  "publishConfig": {
    "directory": "../../dist/libs/api"
  }
}
