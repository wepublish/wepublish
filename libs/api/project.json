{
  "name": "api",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/api/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/api/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/api/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/api",
        "tsConfig": "libs/api/tsconfig.lib.json",
        "packageJson": "libs/api/package.json",
        "main": "libs/api/src/index.ts",
        "assets": ["libs/api/*.md", "libs/api/prisma/**/*"],
        "external": "all",
        "rootDir": "."
      }
    },
    "generate-api": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          {
            "command": "npx graphql-codegen -r dotenv/config --config codegen.yml"
          }
        ]
      }
    }
  },
  "tags": []
}
