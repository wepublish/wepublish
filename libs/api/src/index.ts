export * from './lib/graphql/session'
export * from './lib/graphql/navigation'
export * from './lib/graphql/author'
export * from './lib/graphql/image'
export * from './lib/graphql/blocks'
export * from './lib/graphql/article'
export * from './lib/graphql/comment/comment'
export * from './lib/graphql/page'
export * from './lib/graphql/peer'
export * from './lib/graphql/token'

export * from './lib/graphql/slug'

export * from './lib/graphql/schema'
export * from './lib/graphql/permissions'

export * from './lib/db/user'
export * from './lib/db/userRole'
export * from './lib/db/navigation'
export * from './lib/db/author'
export * from './lib/db/image'
export * from './lib/db/block'
export * from './lib/db/article'
export * from './lib/db/comment'
export * from './lib/db/page'
export * from './lib/db/common'
export * from './lib/db/peer'
export * from './lib/db/memberPlan'
export * from './lib/db/invoice'
export * from './lib/db/payment'
export * from './lib/db/mailLog'
export * from './lib/db/subscription'

export * from './lib/media/karmaMediaAdapter'
export * from './lib/urlAdapter'

export * from './lib/utility'
export * from './lib/error'
export * from './lib/jobs'
export * from './lib/context'
export * from './lib/server'
export * from './lib/payments/paymentProvider'
export * from './lib/payments/stripeCheckoutPaymentProvider'
export * from './lib/payments/stripePaymentProvider'
export * from './lib/payments/payrexxPaymentProvider'

export * from './lib/mails/mailProvider'
export * from './lib/mails/MailgunMailProvider'
export * from './lib/mails/MailchimpMailProvider'
export * from './lib/mails/mailContext'
export * from './lib/challenges/challengeProvider'
export * from './lib/challenges/algebraicCaptchaChallenge'
export * from './lib/payments/payrexxSubscriptionPaymentProvider'

export {getUserForCredentials} from './lib/graphql/user/user.queries'

export * from '@wepublish/settings/api'
export * from '@wepublish/membership/api'
export * from '@wepublish/authentication/api'
export * from '@wepublish/permissions/api'
export * from '@wepublish/consent/api'
export * from '@wepublish/events/import/api'
export * from '@wepublish/image/api'
export * from '@wepublish/richtext/api'
