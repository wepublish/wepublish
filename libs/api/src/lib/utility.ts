import {MemberPlan, PaymentMethod} from '@prisma/client'
import formatISO from 'date-fns/formatISO'
import {GraphQLFieldResolver, GraphQLIsTypeOfFn, GraphQLObjectType} from 'graphql'
import {delegateToSchema, ExecutionResult, IDelegateToSchemaOptions, Transform} from 'graphql-tools'
import {Context} from './context'
import {TeaserStyle} from './db/block'
import {SettingRestriction} from '@wepublish/settings/api'
import {SubscriptionWithRelations} from './db/subscription'
import {UserWithRelations} from './db/user'
import {InvalidSettingValueError} from './error'

export const MAX_PAYLOAD_SIZE = '1MB'

export const ONE_HOUR_IN_MILLISECONDS = 60 * 60 * 1000
export const ONE_DAY_IN_MILLISECONDS = 24 * ONE_HOUR_IN_MILLISECONDS
export const FIFTEEN_MINUTES_IN_MILLISECONDS = 900000
export const ONE_MONTH_IN_MILLISECONDS = 31 * ONE_DAY_IN_MILLISECONDS

export const USER_PROPERTY_LAST_LOGIN_LINK_SEND = '_wepLastLoginLinkSentTimestamp'

export function mapSubscriptionsAsCsv(
  subscriptions: (SubscriptionWithRelations & {
    user: UserWithRelations
    paymentMethod: PaymentMethod
    memberPlan: MemberPlan
  })[]
) {
  let csvStr =
    [
      'id',
      'firstName',
      'name',
      'preferredName',
      'email',
      'active',
      'createdAt',
      'modifiedAt',

      'company',
      'streetAddress',
      'streetAddress2',
      'zipCode',
      'city',
      'country',

      'memberPlan',
      'memberPlanID',
      'paymentPeriodicity',
      'monthlyAmount',
      'autoRenew',
      'startsAt',
      'paidUntil',
      'paymentMethod',
      'paymentMethodID',
      'deactivationDate',
      'deactivationReason'
    ].join(',') + '\n'

  for (const subscription of subscriptions) {
    const user = subscription?.user
    const memberPlan = subscription?.memberPlan
    const paymentMethod = subscription?.paymentMethod
    // if (!user) continue
    csvStr +=
      [
        user?.id,
        `${sanitizeCsvContent(user?.firstName)}`,
        `${sanitizeCsvContent(user?.name)}`,
        `${sanitizeCsvContent(user?.preferredName)}`,
        `${user?.email ?? ''}`,
        user?.active,
        user?.createdAt ? formatISO(user.createdAt, {representation: 'date'}) : '',
        user?.modifiedAt ? formatISO(user.modifiedAt, {representation: 'date'}) : '',
        `${sanitizeCsvContent(user?.address?.company)}`,
        `${sanitizeCsvContent(user?.address?.streetAddress)}`,
        `${sanitizeCsvContent(user?.address?.streetAddress2)}`,
        `${sanitizeCsvContent(user?.address?.zipCode)}`,
        `${sanitizeCsvContent(user?.address?.city)}`,
        `${sanitizeCsvContent(user?.address?.country)}`,
        sanitizeCsvContent(memberPlan?.name),
        subscription?.memberPlanID ?? '',
        subscription?.paymentPeriodicity ?? '',
        subscription?.monthlyAmount ?? '',
        subscription?.autoRenew ?? '',
        subscription?.startsAt ? formatISO(subscription.startsAt, {representation: 'date'}) : '',
        subscription?.paidUntil
          ? formatISO(subscription.paidUntil, {representation: 'date'})
          : 'no pay',
        sanitizeCsvContent(paymentMethod?.name),
        subscription?.paymentMethodID ?? '',
        subscription?.deactivation
          ? formatISO(subscription.deactivation.date, {representation: 'date'})
          : '',
        subscription?.deactivation?.reason ?? ''
      ].join(',') + '\r\n'
  }

  return csvStr
}

/**
 * according to rfc 4180
 * https://www.ietf.org/rfc/rfc4180.txt
 * @param input
 */
function sanitizeCsvContent(input: string | undefined | null) {
  // according rfc 4180 2.7.
  const escapeDoubleQuotes = (input || '').toString().replace(/[#"]/g, '""')
  // according rfc 4180 2.5. / 2.6.
  return `"${escapeDoubleQuotes}"`
}

// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
    .replace(/[ÇĆĈČ]/gi, 'c')
    .replace(/[ÐĎĐÞ]/gi, 'd')
    .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
    .replace(/[ĜĞĢǴ]/gi, 'g')
    .replace(/[ĤḦ]/gi, 'h')
    .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
    .replace(/[Ĵ]/gi, 'j')
    .replace(/[Ĳ]/gi, 'ij')
    .replace(/[Ķ]/gi, 'k')
    .replace(/[ĹĻĽŁ]/gi, 'l')
    .replace(/[Ḿ]/gi, 'm')
    .replace(/[ÑŃŅŇ]/gi, 'n')
    .replace(/[ÒÓÔÕÖØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
    .replace(/[Œ]/gi, 'oe')
    .replace(/[ṕ]/gi, 'p')
    .replace(/[ŔŖŘ]/gi, 'r')
    .replace(/[ßŚŜŞŠ]/gi, 's')
    .replace(/[ŢŤ]/gi, 't')
    .replace(/[ÙÚÛÜŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
    .replace(/[ẂŴẀẄ]/gi, 'w')
    .replace(/[ẍ]/gi, 'x')
    .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
    .replace(/[ŹŻŽ]/gi, 'z')
    .replace(/[·/_,:;\\']/gi, '-')
    .replace(/\s+/g, '-')
    .replace(/[^\w-]+/g, '')
    .replace(/--+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '')
}

export function base64Encode(str: string): string {
  return Buffer.from(str).toString('base64')
}

export function base64Decode(str: string): string {
  return Buffer.from(str, 'base64').toString()
}

export const IsProxiedSymbol = Symbol('isProxied')

export function markResultAsProxied(result: any) {
  if (!result) return null

  for (const key in result) {
    const value = result[key]

    if (typeof value === 'object' && value != null) {
      markResultAsProxied(value)
    }
  }

  return Object.assign(result, {[IsProxiedSymbol]: true})
}

export function isSourceProxied<T>(source: T): source is T & {__typename?: string} {
  return (source as any)[IsProxiedSymbol] == true
}

export function createProxyingResolver<TSource, TContext, TArgs = {[argName: string]: any}>(
  resolver: GraphQLFieldResolver<TSource, TContext, TArgs>
): GraphQLFieldResolver<TSource, TContext, TArgs> {
  return (source, args, context, info) => {
    if (isSourceProxied(source)) {
      return (source as any)[info.path.key]
    }

    return resolver(source, args, context, info)
  }
}

export function createProxyingIsTypeOf<TSource, TContext>(
  isTypeOf: GraphQLIsTypeOfFn<TSource, TContext>
): GraphQLIsTypeOfFn<TSource, TContext> {
  return function (this: GraphQLObjectType, source, context, info) {
    return isSourceProxied(source)
      ? source.__typename === this.name
      : isTypeOf(source, context, info)
  }
}

export function mapEnumsBack(result: any) {
  if (!result) return null

  for (const key in result) {
    const value = result[key]

    if (typeof value === 'object' && value !== null) {
      mapEnumsBack(value)
    }
  }
  if (
    result.__typename === 'ArticleTeaser' ||
    result.__typename === 'PeerArticleTeaser' ||
    result.__typename === 'PageTeaser' ||
    result.__typename === 'EventTeaser' ||
    result.__typename === 'CustomTeaser'
  ) {
    switch (result.style) {
      case 'DEFAULT':
        return Object.assign(result, {style: TeaserStyle.Default})
      case 'LIGHT':
        return Object.assign(result, {style: TeaserStyle.Light})
      case 'TEXT':
        return Object.assign(result, {style: TeaserStyle.Text})
    }
  }
  return result
}

class ResetGraphQLEnums implements Transform {
  transformResult(result: ExecutionResult) {
    // FIXME: WPC-415 created
    return mapEnumsBack(result)
  }
}

export async function delegateToPeerSchema(
  peerID: string,
  fetchAdminEndpoint: boolean,
  context: Context,
  opts: Omit<IDelegateToSchemaOptions, 'schema'>
) {
  const schema = fetchAdminEndpoint
    ? await context.loaders.peerAdminSchema.load(peerID)
    : await context.loaders.peerSchema.load(peerID)

  if (!schema) return null

  return markResultAsProxied(
    await delegateToSchema({
      ...opts,
      schema,
      transforms: [new ResetGraphQLEnums(), ...(opts.transforms ?? [])]
    })
  )
}

export function capitalizeFirstLetter(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function countRichtextChars(blocksCharLength: number, nodes: any) {
  return nodes.reduce((charLength: number, node: any) => {
    if (!node.text && !node.children) return charLength
    if (node.text) {
      return charLength + (node.text as string).length
    }
    return countRichtextChars(charLength, node.children)
  }, blocksCharLength)
}

export type DiscriminateUnion<T, K extends keyof T, V extends T[K]> = T extends Record<K, V>
  ? T
  : never

export type MapDiscriminatedUnion<T extends Record<K, string>, K extends keyof T> = {
  [V in T[K]]: DiscriminateUnion<T, K, V>
}

export function isObject<T>(unknown: unknown): unknown is Record<any, T> {
  return typeof unknown === 'object' && unknown != null && !Array.isArray(unknown)
}

export function isArray<T>(unknown: unknown): unknown is T[] {
  return Array.isArray(unknown)
}

export function isString(unknown: unknown): unknown is string {
  return typeof unknown === 'string'
}

export function isBoolean(unknown: unknown): unknown is boolean {
  return typeof unknown === 'boolean'
}

export function checkSettingRestrictions(
  val: unknown,
  currentVal: unknown,
  restriction: SettingRestriction | undefined
) {
  if (!restriction) {
    return
  }

  if (typeof val !== typeof currentVal) {
    throw new InvalidSettingValueError()
  }

  if (restriction.allowedValues?.boolChoice && typeof val !== 'boolean') {
    throw new InvalidSettingValueError()
  }

  if (typeof val === 'number') {
    if (restriction.maxValue && val > restriction.maxValue) {
      throw new InvalidSettingValueError()
    }

    if (restriction.minValue && val < restriction.minValue) {
      throw new InvalidSettingValueError()
    }
  }

  if (typeof val === 'string') {
    if (restriction.inputLength && val.length > restriction.inputLength) {
      throw new InvalidSettingValueError()
    }

    if (
      restriction.allowedValues?.stringChoice &&
      !restriction.allowedValues.stringChoice.includes(val)
    ) {
      throw new InvalidSettingValueError()
    }
  }
}
