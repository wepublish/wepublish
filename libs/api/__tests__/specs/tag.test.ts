import {CreateTag, DeleteTag, TagList, TagType} from '../api/private'

import {ApolloServer} from 'apollo-server-express'
import {UpdateTag} from '../api/private/index'
import {createGraphQLTestClientWithPrisma, generateRandomString} from '../utility'

let testClientPrivate: ApolloServer
let testClientPublic: ApolloServer

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithPrisma()
    testClientPrivate = setupClient.testServerPrivate
    testClientPublic = setupClient.testServerPublic
  } catch (error) {
    console.log('Error', error)
    throw new Error('Error during test setup')
  }
})

describe('Tags', () => {
  test('can be created', async () => {
    const res = await testClientPrivate.executeOperation({
      query: CreateTag,
      variables: {
        tag: generateRandomString(),
        type: TagType.Comment
      }
    })

    expect(res).toMatchSnapshot({
      data: {
        createTag: {
          id: expect.any(String),
          tag: expect.any(String)
        }
      }
    })
  })

  test('can be updated', async () => {
    const createRes = await testClientPrivate.executeOperation({
      query: CreateTag,
      variables: {
        tag: generateRandomString(),
        type: TagType.Comment
      }
    })

    const tag = generateRandomString()

    const res = await testClientPrivate.executeOperation({
      query: UpdateTag,
      variables: {
        id: createRes.data.createTag.id,
        tag
      }
    })

    expect(res.data.updateTag.tag).toEqual(tag)
  })

  describe('delete', () => {
    test('can be deleted', async () => {
      const createRes = await testClientPrivate.executeOperation({
        query: CreateTag,
        variables: {
          name: generateRandomString(),
          type: TagType.Comment
        }
      })

      const res = await testClientPrivate.executeOperation({
        query: DeleteTag,
        variables: {
          id: createRes.data.createTag.id
        }
      })

      expect(res.data.deleteTag.id).toEqual(createRes.data.createTag.id)
    })
  })

  describe('query', () => {
    describe('private', () => {
      test('can query a list of tags', async () => {
        await Promise.all([
          testClientPrivate.executeOperation({
            query: CreateTag,
            variables: {
              tag: generateRandomString(),
              type: TagType.Comment
            }
          }),
          testClientPrivate.executeOperation({
            query: CreateTag,
            variables: {
              tag: generateRandomString(),
              type: TagType.Event
            }
          })
        ])

        const res = await testClientPrivate.executeOperation({
          query: TagList
        })

        expect(res.data.tags.totalCount).toBeGreaterThanOrEqual(2)
      })
    })

    describe('public', () => {
      test('can query a list of tags', async () => {
        await Promise.all([
          testClientPrivate.executeOperation({
            query: CreateTag,
            variables: {
              tag: generateRandomString(),
              type: TagType.Comment
            }
          }),
          testClientPrivate.executeOperation({
            query: CreateTag,
            variables: {
              tag: generateRandomString(),
              type: TagType.Event
            }
          })
        ])

        const res = await testClientPublic.executeOperation({
          query: TagList
        })

        expect(res.data.tags.totalCount).toBeGreaterThanOrEqual(2)
      })
    })
  })
})
