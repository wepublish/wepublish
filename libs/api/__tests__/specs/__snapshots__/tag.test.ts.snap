// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags can be created 1`] = `
{
  "data": {
    "createTag": {
      "id": Any<String>,
      "tag": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
